docs/architecture/autofill/index.md

```diff
@@ -0,0 +1,123 @@
+| `notification.background.ts` | Handles incoming requests related to the notification bar.                                  |
+| `commands.background.ts`     | Handles incoming requests related to keyboard commands (including autofill).                                    |
+| `contextMenu.background.ts`  | Handles context menu actions (including autofill).                                    |
+| `main.background.js`         | Bootstraps the extension. It is relevant here only because it (arbitrarily) contains the `collectPageDetailsForContentScript()` method. |
+
+In order to support browsers that will require the
+[Chrome Extension Manifest v3](https://developer.chrome.com/docs/extensions/mv3/intro/)
+specifications, the Bitwarden extension also handles the keyboard shortcut command through an
+`onCommandListener`. This listener is registered with `chrome.commands.onCommand.addListener`. The
+`onCommandListener` is responsible for initiating the `AutoFillActiveTabCommand` to trigger the
+Autofill functionality, similar to how the `commands.background.ts` background page handles incoming
+keyboard command messages.
+
+### Messaging
+
+We have established that the Bitwarden extension uses content scripts and background pages or
+listeners in order to perform Autofill. The last piece of the puzzle is the communication between
+them. The Autofill architecture leverages the
+[extension messaging API](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/sendMessage)
+in order to communicate between the extension UI, the background pages, and the content scripts that
+are running on each browser tab.
+
+```kroki type=plantuml
+@startuml
+
+ package "Browser Tab" as T {
+    [Page Source]
+    [Content Scripts]
+ }
+
+ node "Bitwarden Browser Extension" as E {
+    [Browser Extension UI]
+    [Background Pages]
+ }
+
+ [Background Pages] --> [Content Scripts] : tabSendMessage
+ [Content Scripts] --> [Background Pages] : sendMessage
+
+@enduml
+```
+
+#### Sending a request from the content script to the extension
+
+To send a request from a content script to the extension, we have two services provided:
+
+- When Dependency Injection is available, the `BrowserMessagingService` should be used, with its
+  provided `send()` method.
+- When no Dependency Injection is available, the static `BrowserApi` should be used, with its
+  `sendMessage()` method.
+
+On the extension, we attach with `chrome.runtime.onMessage.addListener` to receive the messages
+destined for the extension.
+
+:::note
+
+The `BrowserMessagingService` and `BrowserApi` abstract the `chrome.runtime.sendMessage` API. If any
+direct references to this API are found, they should be refactored to use one of our abstractions.
+
+:::
+
+#### Sending a request from the extension to the content script
+
+We use `BrowserApi.tabSendMessage` to send messages from the extension to the content script running
+on the browser tab.
```

Comment:
I don't think we use the abstraction in content scripts, probably because most of them aren't transpiled, they're just straight JS. Something I'd like to fix though.

Reply:
I think this is right then, isn't it?  We use `BrowserApi.tabSendMessage` (the abstraction) in the extension background page, and receive it with `chrome.runtime.onMessage.addListener` on the content script itself.  Was this comment intended for me to create a note to that effect in the docs, to explain why?

Reply:
Sorry, I misread this and mixed up who the sender was. All good